### ISSUE TITLE
[no-unsafe-return] Error although return type is declared

### ISSUE THREAD

**Issue by screendriver** [typescript-eslint/typescript-eslint#2118]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unsafe-return": "error"
  }
}
```

```TS
export function parse(body: string): MyBody {
  return JSON.parse(body);
}
```

**Expected Result**

Should work as with v2

**Actual Result**

`error  Unsafe return of an any typed value  @typescript-eslint/no-unsafe-return`

Only works with explicit type assertion

```TS
export function parse(body: string): MyBody {
  return JSON.parse(body) as MyBody;
}
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.0.2` |
| `@typescript-eslint/parser`        | `3.0.2` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `14.2.0` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
This is working as expected. The rule detects when you are returning `any` from a function with a type annotation.

Returning `any` is a huge safety hole, doubly so with an explicit function return type.

If you really, really, really are okay with this safety hole, you should use a disable comment.  
I would avoid using an explicit type assertion.

---

For example, say something goes wrong somewhere and your body string is instead `"1"`. That string gets passed into your `parse` function and it simply returns `1`. And now your application breaks in potentially weird ways because of assumptions you (rightly) made based on an unsafe return.

Your types explicitly declared that the `parse` function will always return a `MyBody` type, but it's very possible for it to return a string, boolean, number, null, an array, or an object.

**Comment by diegovincent** [comment]
@bradzacher but what is the alternative? If they do not use the explicit type assertion, the linter will say that there is an unsafe `any` assignment, and `JSON.parse` by definition returns any. What is the correct play here?

**Comment by bbarry** [comment]
In this case: an explicit cast.

`JSON.parse` is a textbook example of an unsafe call to make from a type perspective. You could always actually write a safe parser that explicitly produced your type or you could have a bunch of checks after calling `JSON.parse` to make sure you get what you are really wanting.

**Comment by bradzacher** [comment]
@bbarry is correct here.

The safest code would be something like this, which parses the json, and then validates that the parsed object looks like you expect.

```ts
interface MyData {
    prop1: string;
    prop2?: number;
    prop3?: boolean | null;
}

// ideally we wouldn't need this, but we do due to https://github.com/microsoft/TypeScript/issues/21732
function isObject(thing: unknown): thing is Record<string, unknown> {
    return thing && typeof thing === 'object';
}

function parseMyData(json: string): MyData {
    const raw = JSON.parse(json) as unknown;
  
    if (!isObject(raw)) {
        throw new Error('unexpected JSON format');
    }

    if (!('prop1' in raw) && typeof raw.prop1 !== 'string') {
        throw new Error('unexpected JSON format');
    }

    if ('prop2' in raw && typeof raw.prop2 !== 'number') {
        throw new Error('unexpected JSON format');
    }

    if ('prop3' in raw && typeof raw.prop3 !== 'number' && raw.prop3 !== null) {
        throw new Error('unexpected JSON format');
    }

    return raw as unknown as MyData;
}
```

_However_; I accept that you can't code this up for every use case, and there's no easy way to create some generic code that validate that a json string matches the expected type, because types are erased at compile time.

There are libraries which let you convert types to JSONSchema, but this is sub-optimal for a lot of cases because JSONSchema isn't the fastest way to validate something.  
It's also sub-optimal because in a lot of cases, your json string is coming from a so-called "trusted source" (like an API response), so you should be able to "trust" that the string is in your expected format, and only do minimal validation, like:
```ts
if (raw && typeof raw === 'object') {
  return raw;
}
throw new Error(`Expected JSON to parse as an object, but got "${json}"`);
```

But alas, I don't know anything about your code or architecture, so I have to comment on the general case of parsing json being super unsafe.

If you're okay with returning an unsafe any in a specific case because, for example, your json string is fully "trusted" - then feel free to use a disable comment to disable the rule:

```ts
export function parse(body: string): MyBody {
  // We are assuming that the json string is trusted, and in the correct format
  // eslint-disable-next-line @typescipt-eslint/no-unsafe-return
  return JSON.parse(body);
}
```

**Comment by diegovincent** [comment]
That is what I had in mind, it's nice to see validation and response on your behalf @bradzacher you get my star! Although, simple proposal, should we put something like this in the docs? I can help with that.

**Comment by bradzacher** [comment]
It might be good as an FAQ answer, something like "I get errors from a `no-unsafe-*` rule when using `JSON.parse`, what can I do?"

**Comment by diegovincent** [comment]
Where can I do the PR? Glad to help.

**Comment by bradzacher** [comment]
This is the FAQ file:
https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md
